src/self_driving/pilot_util.cpp

```diff
@@ -36,49 +166,45 @@ const std::list<metrics::PipelineMetricRawData::PipelineData> &PilotUtil::Collec
   execution::exec::OutputCallback callback = consumer;
 
-  execution::query_id_t qid;
   catalog::db_oid_t db_oid;
-  for (auto &it : forecast->query_id_to_params_) {
-    qid = it.first;
-    for (auto &params : forecast->query_id_to_params_[qid]) {
-      txn = txn_manager->BeginTransaction();
-      auto stmt_list = parser::PostgresParser::BuildParseTree(forecast->query_id_to_text_[qid]);
-      db_oid = static_cast<catalog::db_oid_t>(forecast->query_id_to_dboid_[qid]);
-      std::unique_ptr<catalog::CatalogAccessor> accessor =
-          catalog->GetAccessor(common::ManagedPointer(txn), db_oid, DISABLED);
 
-      auto binder = binder::BindNodeVisitor(common::ManagedPointer(accessor), db_oid);
-      binder.BindNameToNode(common::ManagedPointer(stmt_list), common::ManagedPointer(&params),
-                            common::ManagedPointer(&(forecast->query_id_to_param_types_[qid])));
+  std::unordered_set<execution::query_id_t> qids;
+  for (auto i = start_segment_index; i <= end_segment_index; i++) {
+    for (auto &it : forecast->forecast_segments_[i].id_to_num_exec_) {
+      qids.insert(it.first);
+    }
+  }
 
-      // Creating exec_ctx
-      std::unique_ptr<optimizer::AbstractCostModel> cost_model = std::make_unique<optimizer::TrivialCostModel>();
+  for (const auto &qid : qids) {
+    txn = txn_manager->BeginTransaction();
+    auto stmt_list = parser::PostgresParser::BuildParseTree(forecast->query_id_to_text_[qid]);
+    db_oid = static_cast<catalog::db_oid_t>(forecast->query_id_to_dboid_[qid]);
+    std::unique_ptr<catalog::CatalogAccessor> accessor =
+        catalog->GetAccessor(common::ManagedPointer(txn), db_oid, DISABLED);
 
-      auto out_plan =
-          trafficcop::TrafficCopUtil::Optimize(common::ManagedPointer(txn), common::ManagedPointer(accessor),
-                                               common::ManagedPointer(stmt_list), db_oid, pilot->stats_storage_,
-                                               std::move(cost_model), forecast->optimizer_timeout_)
-              ->TakePlanNodeOwnership();
+    auto out_plan = PilotUtil::GetOutPlan(
+        txn, common::ManagedPointer(accessor), common::ManagedPointer(&(forecast->query_id_to_params_[qid][0])),
+        common::ManagedPointer(&(forecast->query_id_to_param_types_[qid])), common::ManagedPointer(stmt_list), db_oid,
+        pilot->stats_storage_, forecast->optimizer_timeout_);
 
+    execution::compiler::ExecutableQuery::query_identifier.store(qid);
+    auto exec_query = execution::compiler::CompilationContext::Compile(*out_plan, exec_settings, accessor.get(),
+                                                                       execution::compiler::CompilationMode::OneShot);
+    pipeline_qids->push_back(qid);
+    for (auto &params : forecast->query_id_to_params_[qid]) {
       auto exec_ctx = std::make_unique<execution::exec::ExecutionContext>(
           db_oid, common::ManagedPointer(txn), callback, out_plan->GetOutputSchema().Get(),
```

Comment:
Actually, we need to create a new transaction for each query and then abort the transaction right away as well. Forgot to mention that before. The `ExecutableQuery` can be reused across transactions. But the original transaction that creates the `ExecutableQuery` also needs to be aborted.